### ISSUE TITLE
Rexray causes AWS EBS inaccessible?

### ISSUE THREAD

**Issue by leokhoa** [codedellemc/rexray#427]
Hi,
Attaching & Detaching EBS using rexray may cause EBS inaccessible.
Steps:

1)  Create a volume:  

```
docker volume create --driver=rexray --name=postgresql --opt=size=5
```

2)  Run docker on a node: nodeA

```
docker run -d -e POSTGRES_PASSWORD=mysecretpassword --volume-driver=rexray   --name hhpg  -v postgresql:/var/lib/postgresql/data postgres 
```

Result ----------> Volume postgresql will be attached with nodeA

3) Run docker on another node: nodeB (Volume postgresql is still attached with nodeA)

```
 docker run -d -e POSTGRES_PASSWORD=mysecretpassword --volume-driver=rexray   --name hhpg  -v postgresql:/var/lib/postgresql/data postgres 
```

Result ----------> Volume postgresql will be auto detached from nodeA and attached to nodeB

---

The actions above seems fine until I repeat action 2 & 3 some more times and suddenly the postgresql volume is not accessible!!!!
I also have cases that another volume created by rexray not accessible (show attaching status forever).
rexray is cool, so please fix the issue.

I'm using this rexray version:

```
SemVer: 0.3.3
OsArch: Linux-x86_64
Branch: v0.3.3
Commit: b30fb870b5b94cd8368824460ad020bfcf20be3a
Formed: Thu, 21 Apr 2016 19:41:36 UTC
```

Another infos:

```
OS:  CentOS Linux release 7.2.1511
Docker: 1.11.1
Swarm: swarm/1.2.1
AMI: ami-c74789a9 (CentOS Linux 7 x86_64 HVM EBS 1602-b7ee8a69-ee97-4a49-9e68-afaee216db2e-ami-d7e1d2bd.3)
Environment: AWS Seoul region
```

**Comment by kacole2** [comment]
@leokhoa i'll look into this for you. Thanks for sharing your reproduction steps. Is there a certain public AMI you used?

**Comment by clintonskitson** [comment]
@leokhoa Thank you for the information.

I believe you are running into issues relation to your expectations around `pre-emption`. The expectation for this feature today is that a container scheduler has decided that a container must be started elsewhere and it has no ability to contact the host to stop the container and thus detach the volume (host down). 

What you are describing look to relate to existing bugs in Ubuntu+EC2 that we have seen where the driver package in Ubuntu for EBS volumes has problems with forceful detaches. Once a forceful detach occurs it can cause existing devices to not be able to be discovered and made available on the OS. I see you said CentOS here, is this the case for all of your testing?

Does this make sense? If pre-emption is enabled, the expectation is generally that the host would *not* be using volumes until it is made available again or rebooted. This expectation would negate what you are seeing with the volumes. However, the OS *should* be able to handle the forceful detaches/re-attaches gracefully to make the point mute. We can do more verification on your CentOS rev to see about replicating the problem.

**Comment by kacole2** [comment]
@clintonskitson i'm noticing similar behavior. The first failover works without issue. However, subsequent failovers are not working. 

I have restarted the REX-Ray service on each host and also `rexray volume detach --volumeid=XYX` to verify it's at an "available" state within AWS and not tied to a single host. When trying to recreate the container again with 
```
docker run -d -e POSTGRES_PASSWORD=mysecretpassword --volume-driver=rexray   --name hhpg  -v postgresql:/var/lib/postgresql/data postgres 
```

Debug output:
```
[centos@awsdev2 rexray]$ tail -f rexray.log
time="2016-05-11T17:36:51Z" level=info msg="os driver initialized" moduleName=default-docker provider=linux
time="2016-05-11T17:36:51Z" level=debug msg="checking volume path cache setting" pathCache=true
time="2016-05-11T17:36:51Z" level=info msg=vdm.List driverName=docker moduleName=default-docker
time="2016-05-11T17:36:51Z" level=info msg="listing volumes" driverName=docker moduleName=default-docker
time="2016-05-11T17:36:52Z" level=info msg=sdm.GetVolume driverName=ec2 moduleName=default-docker volumeID= volumeName=
time="2016-05-11T17:36:52Z" level=info msg=odm.GetMounts deviceName= driverName=linux moduleName=default-docker mountPoint=
time="2016-05-11T17:36:52Z" level=debug msg="docker voldriver spec file" path="/etc/docker/plugins/rexray.spec"
time="2016-05-11T17:36:52Z" level=info msg="started module" address="unix:///run/docker/plugins/rexray.sock" name=default-docker typeName=docker
time="2016-05-11T17:36:52Z" level=info msg="service sent registered modules start signals"
time="2016-05-11T17:36:52Z" level=info msg="service successfully initialized, waiting on stop signal"
time="2016-05-11T17:38:12Z" level=info msg=vdm.Path driverName=docker moduleName=default-docker volumeID= volumeName=postgresql
time="2016-05-11T17:38:12Z" level=debug msg="skipping path lookup" driverName=docker moduleName=default-docker volumeID= volumeName=postgresql
time="2016-05-11T17:38:12Z" level=info msg=vdm.Mount driverName=docker moduleName=default-docker newFsType= overwriteFs=false preempt=false volumeID= volumeName=postgresql
time="2016-05-11T17:38:12Z" level=info msg="mounting volume" driverName=docker moduleName=default-docker newFsType= overwriteFs=false volumeID= volumeName=postgresql
time="2016-05-11T17:38:12Z" level=info msg=sdm.GetVolume driverName=ec2 moduleName=default-docker volumeID= volumeName=postgresql
time="2016-05-11T17:38:13Z" level=info msg=sdm.GetVolumeAttach driverName=ec2 instanceID=i-005a789d moduleName=default-docker volumeID=vol-7bb1a2aa
time="2016-05-11T17:38:13Z" level=debug msg="performing precautionary unmount"
time="2016-05-11T17:38:13Z" level=info msg=odm.Unmount driverName=linux moduleName=default-docker mountPoint="/var/lib/rexray/volumes/postgresql"
time="2016-05-11T17:38:13Z" level=info msg=sdm.AttachVolume driverName=ec2 force=true instanceID=i-005a789d moduleName=default-docker runAsync=false volumeID=vol-7bb1a2aa
time="2016-05-11T17:38:13Z" level=info msg="got next device name" driverName=ec2 moduleName=default-docker nextDeviceName="/dev/xvdb"
time="2016-05-11T17:38:13Z" level=info msg="waiting for volume attachment to complete" driverName=ec2 force=true instanceID=i-005a789d moduleName=default-docker runAsync=false volumeID=vol-7bb1a2aa
```

the docker client hangs. Meanwhile within AWS, the EBS volume is "spinning" and acts like it's attaching to the host. However, the container never starts. I have to stop the rexray service and the `docker run` command finishes with the following mount error:
```
docker: Error response from daemon: Post http://%2Frun%2Fdocker%2Fplugins%2Frexray.sock/VolumeDriver.Mount: EOF.
```

Steps to reproduce:

1. Create 2 machines with Docker-Machine using ami-6d1c2007
2. Here is an example of `config.yml`

    ```
    rexray:
      logLevel: warn
      storageDrivers:
      - ec2
      volume:
        mount:
          preempt: true
        unmount:
          ignoreUsedCount: true
    aws:
      accessKey: mykey
      secretKey: mysecret
    ```

3. follow @leokhoa steps for creating the volume and then performing failovers

**Comment by clintonskitson** [comment]
@kacole2 I am assuming that a reboot on the instances will subsequently clear up the behavior. Can you try an earlier AMI than CentOS 7.2?

**Comment by kacole2** [comment]
@clintonskitson correct. A reboot of the host clears up the behavior. I'll add this to the documentation so it's more clear

**Comment by leokhoa** [comment]
@kacole2: I've updated the AMI : ami-c74789a9
You're right, subsequent failovers will not work and make the EBS spinning forever. 
Reboot the instances helps, but I expect timeout and no need to reboot instances as I'm using Docker Swarm cluster.
Here is support message from AWS team:

> Thanks for letting us know that a reboot fixed your issue. 
We've taken a look and it looks like you may not have unmounted that volume before detaching it. This would have meant that the kernel for your OS would have though that the device would still be connected. When you tried to attach it again, your instance would have been rejecting it thinking that it already has a device of that name. 

> If you encounter this issue again, please try attaching the disk as a different device name. If that works, I'd recommend checking that the device was cleanly unmounted prior to be attached. 

> If you have any further questions, please let us know.

I had checked with df -h and seen the mount point disappeared. I had tried to mount with another device name but no luck. However, I hope the message is helpful to you.

**Comment by clintonskitson** [comment]
@leokhoa Thank you for checking with AWS on this. The scenario should be reproduceable without RR where they can verify the bug on their end.  I would succinctly present the workflow as the following.

1) Attach volume on Host #1 
2) Mount volume on Host #1 
3) Do a forceful detach from the AWS console of the volume
4) Attach volume to Host #2 
5) Mount volume on Host #2 
4) Do a forceful detach from the AWS console of the volume
5) Do an unmount on Host #1 for the previous mount path (RR does this, "precautionary unmount")
6) Attach volume

Assuming it will spin at this point based on what you said. Step #5 which is what they were suggesting is already done before the attach occurs.

```
		ctx.Debug("performing precautionary unmount")
		_ = client.OS().Unmount(ctx, mp, opts.Opts)

		var token string
		vol, token, err = client.Storage().VolumeAttach(
			ctx, vol.ID, &types.VolumeAttachOpts{
				Force: opts.Preempt,
				Opts:  utils.NewStore(),
			})
		if err != nil {
			return "", nil, err
		}
```

**Comment by fritz-fritz** [comment]
Encountering the same behavior and logs. Force detaching the volume from the AWS console makes the host responsive again. My REX-ray container is ubuntu... is there a known working OS to use?

Host: AWS EC2 RancherOS AMI v0.4.5
REX-ray Container OS: phusion\baseimage (ubuntu:14.04)
Docker: 1.10.3

What's the resolution to this? Just reboot the host?

**Comment by kacole2** [comment]
@fritz-fritz The documentation calls out no Ubuntu support in EC2 for the host OS. Please read this comment: https://github.com/emccode/rexray/issues/427#issuecomment-218500005

CentOS, CoreOS, Fedora, etc all work in EC2 as the host OS.

There should not be anything limiting from the baseimage OS.

**Comment by fritz-fritz** [comment]
@kacole2 If this is the case perhaps the docs should be clarified. My interpretation was that Ubuntu is supported but preemption does not. `baseimage` is an optimized image of ubuntu and so any bugs would likely carry over.

I am running a containerized OS with REX-ray within a privileged container. The volume driver is recognized by docker upstream of the container. Rebooting seems to have resolved the problem but isn't a great solution. Created a new issue #455 for possible underlying problem?

**Comment by kacole2** [comment]
@fritz-fritz correct. No pre-emption support for Ubuntu on EC2. Sorry, misread your comment
